design-proposals/crd-based-management.md

```diff
@@ -0,0 +1,417 @@
+
+#### Peer CRD
+
+The `Peer` object defines a remote cluster and its ClusterLink gateways. The local
+ peer is not represented by a dedicated `Peer` CRD instance, but rather through a
+ collection of other k8s objects such as deployment, secret, etc.
+
+```go
+type Peer struct {
+  metav1.TypeMeta  
+  metav1.ObjectMeta // Peer name must match the Subject name presented in its certificate
+
+  Spec   PeerSpec
+  Status PeerStatus
+}
+
+type PeerSpec struct {
+  Gateways []string // one or more gateway addresses, each in the form "host:port"
+  Attributes map[string]string // Peer's attribute set
+  // TODO: should we have fixed/required set of attributes explicitly called out and
+  // an optional attribute set encoded as a map?)
+}
+ 
+type PeerStatus struct {
+  ObservedGeneration int64
+  Conditions[] metav1.Condition
+}
+```
+
+Peers are added to the ClusterLink namespace by the site administrator. Information
+ regarding peer gateways and attributes is communicated out of band and not in scope
+ for this design. Not having any gateways is an error but other than that there is
+ no actual state for a Peer and the object can be reconciled immediately. Besides the
+ recommended reconciliation condition types, a Peer should also support a
+ `Reachable` (or `Seen`) condition indicating whether the peer is currently reachable,
+ and the last time it successfully responded to heartbeats.
+
+Peer names are unique and must align with the Subject name present in their certificate
+ during connection establishment. The name is used by importers in referencing an export.
+
+#### Export and Import
+
+The `Export` object makes a local k8s Service endpoint potentially accessible to other
+ clusters (if allowed if policy). The control plane must not allow access to services
+ not explicitly exported. An `Export` can only refer to a Service in the same namespace
+ as it but the Service type can be any of the supported k8s Services types (specifically
+ allowing `ExternalName` Services). Each export defines a single port and exporting a
+ multi-port Service requires multiple `Export`s defined.
+
+```go
+type Export struct {
+  metav1.TypeMeta  
+  metav1.ObjectMeta
+
+  Spec   ExportSpec
+  Status ExportStatus
+}
+
+type ExportSpec struct {
+  Service string // name of exported service. If omitted, uses the Object name
+                 // from the metadata. The exported Service must be defined and
+                 // in the current namespace
+  Port int16 // exported service port 
+             // TODO: should we support named ports or only numbers?
+             // TODO: Port can be omitted if the Service has a single port?
+}
+
+type ExportStatus struct {
+  ObservedGeneration int64
+  Conditions[] metav1.Condition
+  // TODO: are any additional fields needed? For example, should we reflect the actual port here?
+}
+```
+
+TODO: how and where are Service attributes defined? Both import and export are set
```

Comment:
Why do peers have an attribute field while exports and imports do not?

Reply:
We want to capture Service attributes on the Import/Export. The issue is that these are set by namespace users and Service attributes seem to be more sensitive than to allow arbitrary set by any user (they affect the policies).
Import seems the least likely object for it, Exports are better, but having these in the privileged namespace would be best, IMHO